### ISSUE TITLE
VS Test Retry Has Stopped Working On Azure Pipeline

### ISSUE THREAD

**Issue by bendursley** [microsoft/azure-pipelines-tasks#16990]
We've noticed in the last few days that a couple of our VS Test runs no longer retries on failures, and upon retry, it is unable to find the name spaces for those tests.  
The tests are written in SpecFlow 3.9.74.  We were using an old Test SDK (16.5) but on upgrade to 17.3.2, we still have this issue.  We are running on Windows-Latest (2022). Tried with 2019 but had the same issue as well.  We did not use VS Installer first, but I've tried with that, and we still have the issue where the retries no longer work

We run using the below YAML code:
`      - task: VSTest@2
        displayName: 'VsTest - testAssemblies'
        inputs:
          testAssemblyVer2: |
            **\JourneyTests.dll
            !**\*TestAdapter.dll
            !**\obj\**
          rerunFailedTests: true
          rerunFailedThreshold: 35
          rerunMaxAttempts: 2
          testRunTitle: '${{ parameters.Environment }} - Journey Tests'
          searchFolder: $(PIPELINE.WORKSPACE)\${{ parameters.PipelineResource }}\${{ parameters.ArtifactResource }}
          ${{ if ne(parameters.CategoryFilter, '') }}:
            testFiltercriteria: ${{ parameters.CategoryFilter }}
        env:
          ${{ if eq(parameters.UseMocks, true) }}:
            UseMocks: true
          ${{ else }}:
            UseMocks: false`

In our logs, you can see it start the retry attempt, then fail as it is unable to match the name space

No test matches the given testcase filter `FullyQualifiedName=JourneyTests.Features.HappyPathFeature.New user sign up journey`.

This seems to have changed sometime between Tuesday of this week and Wednesday of this week.  We run these tests on a schedule, so there was no code change between the runs on Tuesday and Wednesday.

Tue: `Microsoft (R) Test Execution Command Line Tool Version 17.3.0-preview-20220626-01 (x64)` `[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.5+1caef2f33e (64-bit .NET Core 3.1.29)` `DTA Execution Host 19.210.32906.4`

Wed: `Microsoft (R) Test Execution Command Line Tool Version 17.3.0-preview-20220626-01 (x64)` `[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.5+1caef2f33e (64-bit .NET Core 3.1.29)` `DTA Execution Host 19.204.32418.1`

## Environment
- Azure Pipelines
- Windows 2022

**Comment by yirkha** [comment]
This is caused by a VSTest task upgrade from 2.205.0 to 2.210.0, done in #16853. It broke handling of tests with a different/custom display name and fully qualified name.

It will not help to update or change the versions of Visual Studio, testing framework or build agents. The VSTest@2/3 tasks need
to be fixed (they reference a newer dependency which broke backwards compatibility).

---

When tests are simply run without extra options, VSTest.Console.exe is executed directly, the test results are published by the agent to the pipeline and that's it.

However, if retries are enabled, a closed source "DtaExecutionHost" component is put in between. It runs VSTest.Console.exe potentially multiple times, parses the output .trx files itself, publishes that to the pipeline, determines the failing ones back through some REST API, handles the reruns, and so on.

This DtaExecutionHost got upgraded from 19.204.32418.1 to 19.210.32906.4 during the upgrade of the VSTest ADO task. The way how a dependency "Microsoft.TeamFoundation.TestClient.PublishTestResults.dll" inside the DtaExecutionHost parses the .trx files changed, and it broke the test reruns for customers.

----

For example, for a parametrized test case with the following definition

```csharp
    [TestMethod]
    [DataTestMethod]
    [DataRow("One")]
    [DataRow("Two")]
    public void TestMethod(string Parameter)
```

The .trx file will have something like this:

```xml
  <TestDefinitions>
    <UnitTest name="TestMethod (One)"
               storage="c:\whatever\bin\debug\net6.0\testproject.dll"
               id="df932f1e-0bdf-4fd8-b4e0-1d20f4f13e98">
      <Execution id="101c7677-781f-43dd-9240-442df2afb88d" />
      <TestMethod codeBase="C:\Whatever\bin\Debug\net6.0\TestProject.dll"
               adapterTypeName="executor://mstestadapter/v2"
               className="Example.Project.Tests.TestClass"
               name="TestMethod" />
    </UnitTest>
```

Note the `<UnitTest name="TestMethod (One)" ...` has spaces and parentheses.
But `<TestMethod className="className="Example.Project.Tests.TestClass" name="TestMethod" ...` does not.

Now the change in DtaExecutionHost:
```diff
--- before\Microsoft.TeamFoundation.TestClient.PublishTestResults-19.204.32418.1\Microsoft.TeamFoundation.TestClient.PublishTestResults\TrxResultParser.cs	Mon Oct 03 16:56:16 2022
+++ after\Microsoft.TeamFoundation.TestClient.PublishTestResults-19.210.32906.4\Microsoft.TeamFoundation.TestClient.PublishTestResults\TrxResultParser.cs	Mon Oct 03 16:55:38 2022
@@ -229,9 +229,9 @@
 			}
 			XmlNode testResultNode = definitionNode.SelectSingleNode("./TestMethod");
 			string automatedTestName = null;
-			if (testResultNode?.Attributes["className"] != null && testResultNode.Attributes["name"] != null)
+			if (testResultNode?.Attributes["className"] != null && (definitionNode.Attributes["name"]?.Value != null || testResultNode.Attributes["name"].Value != null))
 			{
-				string testCaseName = testResultNode.Attributes["name"].Value;
+				string testCaseName = ((definitionNode.Attributes["name"]?.Value == null) ? testResultNode.Attributes["name"].Value : definitionNode.Attributes["name"].Value);
 				string className = testResultNode.Attributes["className"].Value;
 				automatedTestName = ((testCaseName.StartsWith(className) && testCaseName.Contains(".")) ? testCaseName : string.Join(".", className, testCaseName));
 				if (_chutzpahTestFileExtensions.Contains(storageExtension?.Trim(), StringComparer.OrdinalIgnoreCase) && !string.IsNullOrEmpty(storageFileName))
```

This starts overriding the test case name from the `name` attribute on the `<UnitTest/>`.
Before: `automatedTestName = TestMethod["className'"] + "." + TestMethod["name"]`
After: `automatedTestName = TestMethod["className"] + "." + (UnitTest["name"] ?? TestMethod["name"])`

In practice, instead of making the rerun as:
`VSTest.Console.exe /TestCaseFilter:"FullyQualifiedName=Example.Project.Tests.TestClass.TestMethod"`
and unfortunately rerunning all variants of the test, but at least working, it now results in:
`VSTest.Console.exe /TestCaseFilter:"FullyQualifiedName=Example.Project.Tests.TestClass.TestMethod (One)"`
Which does not work at all.

It is wrong for two reasons:
1. The paretheses are not escaped correctly
`Incorrect format for TestCaseFilter Missing Operator '|' or '&'. Specify the correct format and try again. Note that the incorrect format can lead to no test getting executed.`
2. It is not a valid FullyQualifiedName, that needs really just the method FQN. Comparing the display name or whatever would need a completely different filter being made.

For Ben above, it is even simpler - just a test display name with spaces, resulting in
`FullyQualifiedName=JourneyTests.Features.HappyPathFeature.New user sign up journey`
instead of a proper method name. Again resulting in no test being found to retry.

**Comment by yirkha** [comment]
As for a workaround, referencing the task version explicitly in the YAML file works, i.e. instead of just
```
    - task: VSTest@2
# or
    - task: VSTest@3
```
write
```
    - task: VSTest@2.205.0  # Temporary fix for https://github.com/microsoft/azure-pipelines-tasks/issues/16990
# or
    - task: VSTest@3.205.0  # Temporary fix for https://github.com/microsoft/azure-pipelines-tasks/issues/16990
```

**Comment by bendursley** [comment]
@yirkha thank you for the detailed response, it's really insightful!  Didn't realise you could target specific versions beyond the major version.  

We've started to use @2.205.0 and it looks like our retries are now working again, so at least for now we have a workaround.  The TestCaseFilter looks radically different between the two versions of VSTest we use.

**Comment by yirkha** [comment]
Yeah I also didn't know that version pinning worked to the last digit, very fortunate and useful. Thanks for the confirmation from elsewhere!

**Comment by Mortana89** [comment]
Experiencing the same behavior, using Specflow tests, retries fail since a few days, breaking our regression tests.

**Comment by Mortana89** [comment]
> - task: VSTest@2.205.0  # Using a full version temporarily to fix https://github.com/microsoft/azure-pipelines-tasks/issues/16990

This is only possible on YAML CI/CD pipelines though, in case anybody is wondering.
We're still using classic release pipelines, and sadly I don't see  an option to specify this.
Would there be another way around?

**Comment by yirkha** [comment]
@Mortana89 Hmm, maybe you can have luck using the ADO REST API, changing the JSON definition directly. Otherwise just escalate this through your MS support contact, I guess :-/

**Comment by LURKEN** [comment]
Same problem here, all our pipelines just stopped working over the weekend.

![image](https://user-images.githubusercontent.com/623142/193854653-ac9ae751-53cc-41d3-9b1e-d0caffff18b5.png)

**Comment by Mortana89** [comment]
Just for anyone stumbling upon this article, I could get it working with manually setting the batch options to a very high number (e.g. 10000). Only downside is that these tests now take roughly 2-3x as long as normally.

**Comment by MBGoldberg-BD** [comment]
We are also having this issue with our classic pipeline. No other remedy except for VSTest to provide new release.

**Comment by vinayakmsft** [comment]
We are working on this, will update the Vstest task soon

**Comment by fravaee** [comment]
FRavaee any estimate?

**Comment by 2DORSC** [comment]
we are facing the same issue, do you have any updates/estimates?

**Comment by vinayakmsft** [comment]
We have taken this as our work item , maybe by next week we will update

**Comment by swapmali99** [comment]
Any update on this?

**Comment by Mortana89** [comment]
> We have taken this as our work item , maybe by next week we will update

One month later, what is the status here?

**Comment by MBGoldberg-BD** [comment]
I see related bug/issue  4036 was closed because " issue exists on azdo side". See https://github.com/microsoft/vstest/issues/4036
Still waiting for this issue to be fixed.

**Comment by Morten4** [comment]
Also having this problem!

**Comment by DavidDanaj** [comment]
@vinayakmsft Any update? Issue is still present

**Comment by vinayakmsft** [comment]
Sorry for the delay in response. We will raise a PR for the fix, and it would start working after the deployments.

**Comment by fravaee** [comment]
Any update?

**Comment by vinayakmsft** [comment]
Please find the tracking work item for the issue. https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/1995653/

**Comment by yaterchatter** [comment]
The tracking work item linked is not accessible. I get 403 Forbidden after signing up.

**Comment by Mortana89** [comment]
Hi, this is open for almost 6 months.
Can we get a decent statusupdate/ETA please?
This breaking change is impacting workflow a lot.

**Comment by fravaee** [comment]
Hi,

I strongly agree. PLEASE PLEASE

**Comment by MBGoldberg-BD** [comment]
> The tracking work item linked is not accessible. I get 403 Forbidden after signing up.

Likewise. Can't access. Still looking for update on this.

**Comment by yirkha** [comment]
The link goes to a MS-internal ADO instance. Perhaps Vinayak thought that there were mostly colleagues commenting here. No point in trying to access it for anyone who is not a MS employee or does not have the right privileges.

Now, I _can_ open it, but I can't comment on what's inside still, I'm sorry. First because it's confidential, duh, second because you would probably lynch _me_ for it, even though I'd be just a messenger >.<

But I'd say, if you are able to work around this on your side, in any way... do it.

**Comment by ismael-ms** [comment]
Are there any updates on a fix? My team has been experiencing the same issues as well https://github.com/microsoft/azure-pipelines-tasks/issues/18084.

**Comment by eyalLefler** [comment]
The issue is still open, will it be fix ever?
By the way, I'm unable to run the task: - task: VSTest@2.205.0
Getting the error: "A task is missing. The pipeline references a task called 'VSTest'. This usually indicates the task isn't installed..."
Any idea how to fix it?

**Comment by MBGoldberg-BD** [comment]
Here we are August 2023, issue was injected Sep 2022.
Can we please get this fixed?  Give us a status update.
Michael

**Comment by yirkha** [comment]
I've reached out to the engineer who was working on this and asked.

**Comment by yirkha** [comment]
Should be fixed in VSTest@2.225.0 after #18760 gets merged

**Comment by yirkha** [comment]
Thank for the update. The new task version does not seem to have the desired effect on our pipeline.

Before (v2.224.0):
```
2023-08-16T08:49:17.0628419Z Task : Visual Studio Test
2023-08-16T08:49:17.0628738Z Version : 2.224.0 
...
2023-08-16T08:49:17.9948378Z DtaExecutionHost version 19.210.32906.4. 
...
2023-08-16T09:06:29.5685456Z Failed Test_DisconnectOnLargeMessage(v4) [39 s]
...
2023-08-16T09:10:37.8506695Z **************** Rerunning failed tests for Test run 326350859 ********************* 
2023-08-16T09:10:37.8948952Z vstest.console.exe /TestCaseFilter:
        "FullyQualifiedName=Trouter.FunctionalTests.TrouterNativeNetClientFunctionalTest.Test_DisconnectOnLargeMessage(v4)"
...
2023-08-16T09:10:39.6810636Z ##[error]Incorrect format for TestCaseFilter Missing Operator '|' or '&'.
                             Specify the correct format and try again. Note that the incorrect format
                             can lead to no test getting executed. 
```

After (v2.227.0):
```
2023-08-24T19:39:22.9042215Z Task : Visual Studio Test
2023-08-24T19:39:22.9042542Z Version : 2.227.0 
...
2023-08-24T19:39:28.8259497Z DtaExecutionHost version 19.226.34002.2. 
...
2023-08-24T19:54:38.0774500Z Failed Test_MessageStore_TrySyncMode(v3) [13 s] 
2023-08-24T19:54:45.0069644Z Failed Test_MessageStore_TrySyncMode(v4) [6 s] 
...
2023-08-24T20:00:32.8905690Z **************** Rerunning failed tests for Test run 327331990 ********************* 
2023-08-24T20:00:32.9432344Z vstest.console.exe /TestCaseFilter:
        "(FullyQualifiedName=Trouter.FunctionalTests.TrouterNativeNetClientFunctionalTest.Test_MessageStore_TrySyncMode(v4))
        |(FullyQualifiedName=Trouter.FunctionalTests.TrouterNativeNetClientFunctionalTest.Test_MessageStore_TrySyncMode(v3))"
...
2023-08-24T20:00:34.9557818Z ##[error]Incorrect format for TestCaseFilter Missing Operator '|' or '&'.
                             Specify the correct format and try again. Note that the incorrect format
                             can lead to no test getting executed. 
```

**Comment by dogirala** [comment]
Adding the observations from the above error, for dynamic tests or tests with custom displayname, please follow the same naming as standard [DataTestMethod] to resolve the above issue. Something like the below overload should work. Please reach out if issue is still observed

'
       public string GetDisplayName(MethodInfo methodInfo, object[] data)
        {
            if (data != null)
            {
                var testParamString = new StringBuilder();
                testParamString.Append(methodInfo.Name);

                testParamString.Append(' (');
                var methodParameters = methodInfo.GetParameters();
                for (int i = 0; i < methodParameters.Length; i++)
                {
                    testParamString.Append(data[i]);
                    testParamString.Append(',');
                }
                testParamString.Length--;
                testParamString.Append(')');

                return testParamString.ToString();
            }

            return methodInfo.Name;
        }
'

**Comment by yirkha** [comment]
I clarified this with @dogirala and the problem turned out to be in our custom `ITestDataSource::GetDisplayName()` implementation, which was formatting parameterized test names as `MethodName(parameters)` (i.e. without a space), which was different from e.g. the standard `[DataTestMethod]`, which uses `MethodName (parameters)` (i.e. with a space before the opening parenthesis).

After changing the display name to match the usual style, everything works fine and the VSTest task correctly retries only the single relevant test case (by using a combined, correctly escaped filter, e.g. `(FullyQualifiedName=Trouter.FunctionalTests.Test_ReconnectTtl_LongPoll&Name=Test_ReconnectTtl_LongPoll \(v3\))`). Thanks for the fix!

**Comment by Liran-Dobrish** [comment]
HI
I also have this problem with my on-prem ADO 2020 (update 1).
How can I get this version of the task?

**Comment by sablanc-msft** [comment]
We still have this problem in our pipelines with version **2.227**.
We are getting this kind of errors (when the pipeline tries to rerun the test):

_No test matches the given testcase filter `(FullyQualifiedName=Microsoft.Office.Project.CloudPcs.WebApiServiceIntegrationTests.ProjectUpdateWorkTemplateIntegrationTests.**UpdateWorkTemplate_TaskUpdates_WhenWorkHourTemplateChangesTo24Hour_UsesDataverse**)` in D:\a\_work\1\a\Tests\IntegrationTests\Pcs\PcsCalendarsIntegrationTests\bin\Release\Pss.CalendarsIntegrationTests.dll_

and with these attributes on the test:

[DataRow(true, DisplayName = "**UpdateWorkTemplate_TaskUpdates_WhenWorkHourTemplateChangesTo24Hour_UsesDataverse**")]
[DataRow(false, DisplayName = "UpdateWorkTemplate_TaskUpdates_WhenWorkHourTemplateChangesTo24Hour")]
public async Task UpdateWorkTemplate_TaskUpdates_WhenWorkHourTemplateChangesTo24Hour(bool shouldUseDataverseCalendarMethods)

**Comment by sablanc-msft** [comment]
I also noticed that the bug has been closed but the associated PR was abandoned:

[Bug 1995653](https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/1995653): Add handling for ( and ) during TestFilter for VsTest task.

Was the issue fixed with another PR?

**Comment by dogirala** [comment]
The issue is due to the usage of custom displayname for the unitests, for rerun of failed tests currently it is not supported, please follow the standard naming for DataTestMethod to resolve it. WIP for supporting custom displayname, will update with an ETA once I have completed my investigation in this sprint.

**Comment by jaydeboer** [comment]
The bug is also impacting our teams. The version workaround is working, but a fix would certainly be preferred.

**Comment by SeMuell** [comment]
@dogirala we are facing the issue with custom display names (that we have to set because of publishing the results to our QA tool). Is there any ETA?

**Comment by dogirala** [comment]
Hi @SeMuell, I have worked on that in last sprint, I have created a pull request, however it is in review and will be merged sometime this week. So please expect the fix sometime by end of next week. I'll update the vstest version here once deployed. 

Also, couple of queries just for clarification, when you mean issue with custom displayname, you mean the rerun of failed tests is not working with tests have custom displayname, right?. And you are using Mstest adapter?

**Comment by SeMuell** [comment]
@dogirala, that would be awesome if it can be merged soon!

To your question, yes, rerunning failed tests with custom displayname. We are using MSTest as adapter. We are also using DataRows, but I duplicated the tests for now.

Running with the workaround of the customBatchSize does now rerun the failed tests successfully, but the test results are immediately cleaned after the test run (which we need to keep) and therefore this approach does not work for us...
When your fix is published, is it still required to use the customBatchSize?

**Comment by dogirala** [comment]
@SeMuell So if I understand correctly you are facing the same issue as below?
![image](https://github.com/microsoft/azure-pipelines-tasks/assets/100841140/0b1023eb-eb61-4435-93f9-51ad9bd963a7)

If yes, then you wouldn't need to set the custombatchsize, it should work as expected without that.

**Comment by SeMuell** [comment]
@dogirala, without having a DataRow, just using the displayName and without setting the customBatchSize, yes.

**Comment by dogirala** [comment]
can you share an example stack trace and an example unittest method where it is failing. It is just for me to retest in local to be sure

**Comment by SeMuell** [comment]
@dogirala, sure, I can share what I have. I just have System.Debug set to true. How can I get you the full stack trace?

Unit test:
```csharp
  [TestMethod(displayName: "QA-VER-100")]
  public void Dashboard_TestStatus_OptimalWorkflow()
  {
    Assert.Fail();
  }
```

VSTest log from AZDO:
```
2024-03-18T11:28:58.0137912Z **************** Rerunning failed tests for Test run 2762726 *********************
2024-03-18T11:28:58.0138063Z Max attempts: 5; Current attempt: 1; Failed test cases threshold: 60; Failed test cases max limit: 0
2024-03-18T11:28:58.0138273Z [command]C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\Common7\IDE\Extensions\TestPlatform\vstest.console.exe "@C:\agent\_work\_temp\vsbcakukxzz.tmp"
2024-03-18T11:28:58.0840682Z Microsoft (R) Test Execution Command Line Tool Version 17.2.0-preview-20220401-07 (x64)
2024-03-18T11:28:58.0841053Z Copyright (c) Microsoft Corporation.  All rights reserved.
2024-03-18T11:28:58.0908780Z vstest.console.exe /TestCaseFilter:"FullyQualifiedName=Project.DashboardTests.QA-VER-100"
2024-03-18T11:28:58.0909115Z "C:\agent\_work\5\drop\tests\Project\bin\x64\Release\Project.dll"
2024-03-18T11:28:58.0909248Z /Settings:"C:\agent\_work\_temp\ngoheiikwlu.tmp.runsettings"
2024-03-18T11:28:58.0909428Z /Logger:"trx"
2024-03-18T11:28:58.0909534Z /TestAdapterPath:"C:\agent\_work\5\drop"
2024-03-18T11:28:58.0909707Z /Diag:"C:\agent\_work\_temp\Attempt-5_t1qtqz.diag"
2024-03-18T11:28:58.7077122Z Starting test execution, please wait...
2024-03-18T11:28:58.7087507Z Logging Vstest Diagnostics in file: C:\agent\_work\_temp\Attempt-5_t1qtqz.diag
2024-03-18T11:28:59.7806751Z A total of 1 test files matched the specified pattern.
2024-03-18T11:28:59.7914900Z Data collection : Logging DataCollector Diagnostics in file: C:\agent\_work\_temp\Attempt-5_t1qtqz.datacollector.24-03-18_12-28-58_90911_4.diag
2024-03-18T11:28:59.8375231Z Blame: Attaching crash dump utility to process testhost (1108).
2024-03-18T11:29:00.1459234Z Logging TestHost Diagnostics in file: C:\agent\_work\_temp\Attempt-5_t1qtqz.host.24-03-18_12-28-59_78411_4.diag
2024-03-18T11:29:00.3898230Z No test matches the given testcase filter `FullyQualifiedName=Project.DashboardTests.QA-VER-100` in C:\agent\_work\5\drop\tests\Project\bin\x64\Release\Project.dll
2024-03-18T11:29:00.6035571Z Results File: C:\agent\_work\5\a\Project.results.xml
2024-03-18T11:29:00.6411112Z Results File: C:\agent\_work\5\a\Svc-testagent_2024-03-18_12_29_00.trx
2024-03-18T11:29:00.6850548Z ##[debug]Exited vstest.console.exe with code 0.
2024-03-18T11:29:00.6851006Z ##[debug]PERF: ExecuteVsTestPhase.InvokeVSTest: took 2671.1391 ms
2024-03-18T11:29:00.6851615Z Vstest.console.exe exited with code 0.
2024-03-18T11:29:00.6851721Z **************** Completed test execution *********************
```

**Comment by SeMuell** [comment]
@dogirala, I got the diag log for the testhost:
[Attempt-5_o3upje.host.24-03-18_12-52-48_84922_4.log](https://github.com/microsoft/azure-pipelines-tasks/files/14636036/Attempt-5_o3upje.host.24-03-18_12-52-48_84922_4.log)

**Comment by dogirala** [comment]
@SeMuell thanks for the log

**Comment by dogirala** [comment]
@SeMuell, I've tried in local it is working as expected without the customBatchSize. 
![image](https://github.com/microsoft/azure-pipelines-tasks/assets/100841140/06e2b831-996e-4dba-b573-c71e8a04368f), will update here once vstest is deployed

**Comment by dogirala** [comment]
The fix is merged it should be deployed sometime by mid to end of next week in vstest 2.237.0 version. Right now the latest version of vstest is 2.236.0

**Comment by SeMuell** [comment]
@dogirala, nice to hear. How can I enable the feature in dev.azure.com?

**Comment by dogirala** [comment]
@SeMuell the fix is merged in vstest 2.x. If it is dev.azure.com, we will take care of enabling the feature, after vstest deployment is done.

**Comment by SeMuell** [comment]
@dogirala, perfect. I'll check as soon as it is published and let you know.

**Comment by SeMuell** [comment]
@dogirala, it seems to be published now, however, the feature flag seems not existing / enabled at the moment in dev.azure.com

```log
##[warning]Failed to get feature flag TestExecution.EnableReRunForDataDrivenTestsWithDisplayName
##[debug]Processed: ##vso[task.logissue type=warning;]Failed to get feature flag TestExecution.EnableReRunForDataDrivenTestsWithDisplayName
```

**Comment by dogirala** [comment]
Hi @SeMuell, apologies for the delay here, I was out of office past couple of weeks and deployments got delayed. I will start enabling the FF from tomorrow. Can you tell me your hostname so that I can share you the expected ETA?

**Comment by SeMuell** [comment]
Hi @dogirala, no worries. Unfortunately, I do not know the hostname of our Azure DevOps instance, region is West Europe.

**Comment by ciconq** [comment]
Hey @dogirala, not sure if this fix has anything to do with the issue I am facing but currently I am not able to rerun even `non-parametrized` `NUnit` tests(`[Test]`). Last time it was working for me was on Friday(with the example bellow).

```
[Category("All")]
[Category("B")]
public class TestsB : BaseHooks
{ 
    [Test]
    public void TC_01B()
    {
        // do something
    }
}
```

This is my `VSTest@2` task:

```
      - task: VSTest@2
        displayName: 'Test run'
        inputs:
          testSelector: 'testAssemblies'
          testAssemblyVer2: |
            **\*Tests*.dll
            !**\*TestAdapter.dll
            !**\obj\**
          searchFolder: '$(System.DefaultWorkingDirectory)'
          testFiltercriteria: 'TestCategory=${{ parameters.CategoryValue }}'
          runTestsInIsolation: false
          platform: '${{ parameters.buildPlatform }}'
          configuration: '${{ parameters.buildConfiguration }}'
          rerunFailedTests: true
          rerunType: 'basedOnTestFailureCount'
          rerunFailedTestCasesMaxLimit: '50'
          rerunMaxAttempts: '3'
```

What I can see now is that it is looking for this `TestCaseFilter(ClassName and Name)`:

```
Max attempts: 3; Current attempt: 1; Failed test cases threshold: 0; Failed test cases max limit: 50
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\Extensions\TestPlatform\vstest.console.exe "@D:\a\_temp\ari41z4npvo.tmp"
Microsoft (R) Test Execution Command Line Tool Version 17.9.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.
vstest.console.exe /TestCaseFilter:"(ClassName=NUnitCalculator.Tests.TestsB&Name=TC_01B)"
```

And when it was working it was looking for this `TestCaseFilter(FullyQualifiedName)`:

```
Max attempts: 3; Current attempt: 1; Failed test cases threshold: 0; Failed test cases max limit: 50
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\Extensions\TestPlatform\vstest.console.exe "@D:\a\_temp\friwkdnkdhd.tmp"
Microsoft (R) Test Execution Command Line Tool Version 17.9.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.
vstest.console.exe /TestCaseFilter:"(FullyQualifiedName=NUnitCalculator.Tests.TestsB.TC_01B)"
```

Both executions are for VS Test `2.238.2`

Any suggestions on this be really appreciated. 

Thank you for your time.

**Comment by ciconq** [comment]
Hey @dogirala, I've observed this on one of my personal projects and created this example. It is not on `Azure DevOps Server` but `Azure DevOps Services`.

**Comment by dancoupar** [comment]
We've fallen victim to this issue in the last couple of days. Our pipeline, which uses a local agent, started spitting out the "No tests were discovered while attempting to rerun failed tests." error despite us not changing anything. Worryingly, the version of the VSTest@2 task is the same as it was when it was working, suggesting changes have been made to the existing version. This made it difficult to figure out what had changed.

Changing the pipeline to use VSTest@3 resolved the issue for us. Still, not ideal and a lot of time wasted today on this.

**Comment by mh884** [comment]
The issue still exist in 2.240.1

`2024-07-06T01:29:06.7861490Z ##[error]Test Run Failed.
2024-07-06T01:29:06.7872567Z  Total time: 2.7576 Hours
2024-07-06T01:29:06.9003794Z Vstest.console.exe exited with code 1.
2024-07-06T01:29:06.9004211Z **************** Completed test execution *********************
2024-07-06T01:29:06.9138261Z Test results files: C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_tempTestResults\TestQAAdmin_FSTestQA1QAVM_2024-07-05_18_43_41.trx
2024-07-06T01:29:08.0904084Z Created test run: 24571
2024-07-06T01:29:08.0905939Z Publishing test results: 773
2024-07-06T01:29:08.2283146Z Publishing test results to test run '24571'.
2024-07-06T01:29:08.2310527Z TestResults To Publish 773, Test run id:24571
2024-07-06T01:29:08.2405898Z Test results publishing 773, remaining: 0. Test run id: 24571
2024-07-06T01:29:16.1785969Z Published test results: 773
2024-07-06T01:29:16.1792053Z Publishing Attachments: 1
2024-07-06T01:29:16.6797433Z Failed tests: 33; Total tests: 773;
2024-07-06T01:29:16.7775678Z Total failed tests as returned by TCM: 33.
2024-07-06T01:29:16.7783700Z Found 1 filtered test case sources.
2024-07-06T01:29:16.8427013Z **************** Rerunning failed tests for Test run 24571 *********************
2024-07-06T01:29:16.8427619Z Max attempts: 3; Current attempt: 1; Failed test cases threshold: 80; Failed test cases max limit: 0
2024-07-06T01:29:16.8428396Z [command]C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\Extensions\TestPlatform\vstest.console.exe "@C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_temp\a2l3uqnbopt.tmp"
2024-07-06T01:29:16.8931944Z Microsoft (R) Test Execution Command Line Tool Version 16.11.0
2024-07-06T01:29:16.8932675Z Copyright (c) Microsoft Corporation.  All rights reserved.
2024-07-06T01:29:16.8970561Z vstest.console.exe /TestCaseFilter:"(ClassName=TestProj.Z_SingleObject&Name=N_TC_BR_FB_29_02)|(ClassName=TestProj.Z_Lookup&Name=N_TC_BR_Lookup_14)|(ClassName=TestProj.Z_FormDashboard&Name=N_TC_BR_FD_02)|(ClassName=TestProj.Z_FlexibleFormControls&Name=N_TC_BR_FFC_39)|(ClassName=TestProj.Z_ConditionalRules&Name=N_TC_BR_CR_FB_37_38)|(ClassName=TestProj.Z_SaveForLater&Name=N_TC_BR_SFL_10)|(ClassName=TestProj.Z_MultiPage&Name=N_TC_BR_MPQA_30)|(ClassName=TestProj.Z_SmokeTest&Name=N_TC_BR_ST_01_Native)|(ClassName=TestProj.Z_SingleObject&Name=N_TC_BR_FB_33_01)|(ClassName=TestProj.Z_FlexibleFormControls&Name=N_TC_BR_FFC_07)|(ClassName=TestProj.Z_AdminDashboard&Name=N_TC_BR_AD_37_40)|(ClassName=TestProj.Z_StyleEditorCtrls&Name=N_TC_BR_SEC_29)|(ClassName=TestProj.Z_DashboardDataView&Name=N_TC_BR_ADDV_01)|(ClassName=TestProj.Z_MultiPage&Name=N_TC_BR_MPQA_39)|(ClassName=TestProj.Z_StyleEditorCtrls&Name=N_TC_BR_SEC_32)|(ClassName=TestProj.Z_Master_QA&Name=N_TC_BR_Test_Special_08)|(ClassName=TestProj.Z_Lookup&Name=N_TC_BR_Lookup_24)|(ClassName=TestProj.Z_New_CSS_Styles&Name=N_TC_BR_NEWCSS_Error_01)|(ClassName=TestProj.Z_FlexibleFormControls&Name=N_TC_BR_FFC_49)|(ClassName=TestProj.Z_eSignature&Name=N_TC_BR_eSign_12)|(ClassName=TestProj.Z_FlexibleFormControls&Name=N_TC_BR_FFC_28)|(ClassName=TestProj.Z_AutoPrefill&Name=N_TC_BR_AP_06)|(ClassName=TestProj.Z_RuleEditorPagination&Name=N_TC_BR_CR_FB_02_01)|(ClassName=TestProj.Z_SingleObject&Name=N_TC_BR_FB_29_01)|(ClassName=TestProj.Z_ConditionalRules&Name=N_TC_BR_CR_FB_63)|(ClassName=TestProj.Z_ConditionalRules&Name=N_TC_BR_CR_FB_51)|(ClassName=TestProj.Z_eSignature&Name=N_TC_BR_eSign_22)|(ClassName=TestProj.Z_RuleEditorPagination&Name=N_TC_BR_CR_FB_02_06)|(ClassName=TestProj.Z_FormDashboard&Name=N_TC_BR_FD_04)|(ClassName=TestProj.Z_ConditionalRules&Name=N_TC_BR_CR_FB_39)|(ClassName=TestProj.Z_RuleEditorPagination&Name=N_TC_BR_CR_FB_01_02)|(ClassName=TestProj.Z_MultiPage&Name=N_TC_BR_MPQA_34)|(ClassName=TestProj.Z_PicklistManagement&Name=N_TC_BR_PICK_04)"
2024-07-06T01:29:16.8974096Z "C:\Workspace\vsts-agent-win-x64-2.175.2\_work\r1\a\_SFApp Forms - Regression - Build\DropRegressionBuild\Project\TestProj\bin\Debug\TestProj.dll"
2024-07-06T01:29:16.8974719Z /Settings:"C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_temp\lngbmr0khhl.tmp.runsettings"
2024-07-06T01:29:16.8975106Z /InIsolation
2024-07-06T01:29:16.8975313Z /Logger:"trx"
2024-07-06T01:29:16.8975624Z /TestAdapterPath:"C:\Workspace\vsts-agent-win-x64-2.175.2\_work\r1\a"
2024-07-06T01:29:17.3201567Z Starting test execution, please wait...
2024-07-06T01:29:17.4869939Z A total of 1 test files matched the specified pattern.
2024-07-06T01:29:18.4763470Z NUnit Adapter 4.5.0.0: Test execution started
2024-07-06T01:29:18.5127857Z Running selected tests in C:\Workspace\vsts-agent-win-x64-2.175.2\_work\r1\a\_SFApp Forms - Regression - Build\DropRegressionBuild\Project\TestProj\bin\Debug\TestProj.dll
2024-07-06T01:29:19.1204574Z    NUnit3TestExecutor discovered 0 of 0 NUnit test cases using Current Discovery mode, Explicit run
2024-07-06T01:29:19.2765187Z NUnit Adapter 4.5.0.0: Test execution complete
2024-07-06T01:29:19.3983320Z No test matches the given testcase filter `(ClassName=TestProj.Z_SingleObject&Name=N_TC_BR_FB_29_02)|(ClassName=TestProj.Z_Lookup&Name=N_TC_BR_Lookup_14)|(ClassName=TestProj.Z_FormDashboard&Name=N_TC_BR_FD_02)|(ClassName=TestProj.Z_FlexibleFormControls&Name=N_TC_BR_FFC_39)|(ClassName=TestProj.Z_Con...` in C:\Workspace\vsts-agent-win-x64-2.175.2\_work\r1\a\_SFApp Forms - Regression - Build\DropRegressionBuild\Project\TestProj\bin\Debug\TestProj.dll
2024-07-06T01:29:19.5073113Z Results File: C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_tempTestResults\TestQAAdmin_FSTestQA1QAVM_2024-07-05_21_29_19.trx
2024-07-06T01:29:19.5328203Z Vstest.console.exe exited with code 0.
2024-07-06T01:29:19.5328700Z **************** Completed test execution *********************
2024-07-06T01:29:19.5377712Z Test results files: C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_tempTestResults\TestQAAdmin_FSTestQA1QAVM_2024-07-05_21_29_19.trx
2024-07-06T01:29:19.5402694Z No Result Found to Publish 'C:\Workspace\vsts-agent-win-x64-2.175.2\_work\_tempTestResults\TestQAAdmin_FSTestQA1QAVM_2024-07-05_21_29_19.trx'.
2024-07-06T01:29:19.5403181Z Updating test results: 0
2024-07-06T01:29:19.5824051Z Publishing test results to test run '24571'.
2024-07-06T01:29:19.8174325Z TestResults To Publish 0, Test run id:24571
2024-07-06T01:29:19.8174800Z Updated test results: 0
2024-07-06T01:29:19.8175103Z Publishing Attachments: 1
2024-07-06T01:29:19.9454973Z ##[error]Error occurred while publishing test results : System.Exception: No tests were discovered while attempting to rerun failed tests. This is likely to happen if these tests are xunit or nunit datadriven/parameterized tests or tests with custom display names with spaces or special characters which are not supported with rerun. Please reach out to us at https://github.com/microsoft/azure-pipelines-tasks/issues if you are hitting this for tests other than the ones mentioned before.
2024-07-06T01:29:20.0541149Z Execution Result Code 1 is non zero, checking for failed results
2024-07-06T01:29:20.1595447Z Completed TestExecution Model...`

**Comment by ciconq** [comment]
Hey @dogirala, are there any updates on this?

**Comment by dancoupar** [comment]
When we first encountered this problem back in April/May, we updated our pipeline to use VSTest@3 instead of VSTest@2, to work around the issue.

Now the exact same problem has reared its head again with VSTest@3.

To the outside world, it appears as though changes are being made haphazardly without due process, namely making breaking changes to an existing version instead creating a new version.

@dogirala back in May, you said "let me dig into this further and getback on the same.", but since then there has been no update.

This all seems a bit unprofessional. Could we have some clarity please?

**Comment by dancoupar** [comment]
Tagging @virekumar @bhaveshbhati @vmapetr @DmitriiBobreshev as you've made commits in the last month. Would be grateful if you could take a look at this issue.

**Comment by virekumar** [comment]
I will have a look at this issue and reply back by with actions planned for it in upcoming week.

**Comment by virekumar** [comment]
We have picked this bug for fix for upcoming sprint. I will inform once this is fixed.

**Comment by dancoupar** [comment]
Hi @virekumar, just wondered where you'd got to with this. Is it still on the backlog?

**Comment by virekumar** [comment]
Tagging manos@microsoft.com

**Comment by freeus-sam** [comment]
@virekumar manos@mircrosoft.com any updates? My pipeline has broken as well

**Comment by csrowell** [comment]
@virekumar [manos@mircrosoft.com](mailto:manos@mircrosoft.com) any updates?

**Comment by virekumar** [comment]
@skmanoj Can you please provide updates here ?

**Comment by sivak1982** [comment]
Any update on this issue?
I am getting this error both on latest VSTest@2 & 3

**Comment by skmanoj** [comment]
Apologies for having this issue in the backlog for a while, due to high volume of incoming issues we couldn't get on this.
we consider this as priority issue which needs to be fixed soon. We will get back on this.

**Comment by urnotdfs** [comment]
I got the same issue on latest VSTest@3. Hope it could be fixed soon.

**Comment by alexButcherMP** [comment]
@Najdenovskii 
Any update on this?

**Comment by Najdenovskii** [comment]
@alexButcherMP yeah Bhavshuk is looking at it

<img width="784" alt="Image" src="https://github.com/user-attachments/assets/2c18b091-c814-4e38-a3f5-ae44411e9a9a" />

**Comment by TimLovellSmith** [comment]
@Najdenovskii Is anyone still looking at this? I can't believe nobody has fixed this for 3 years, when it is the kind of quality issue for the test runners, which can cause pipeline build pass rates to suddenly plummet from 100% to 0%, for no obvious reason, just because someone checked in a particular new flakey test.

**Comment by DarrylGwinnett** [comment]
> The bug will be picked in the current sprint that ends on June. Follow: https://dev.azure.com/mseng/AzureDevOps/_workitems/edit/2215448

Must be long sprints at Microsoft.

We ended up giving up on VSTest and writing our own in-house implementation. 

You should just abandon this product, your support for it is absolutely embarrassing.

**Comment by skmanoj** [comment]
We have already documented the known limitations of the VSTest task in handling data-driven mstest, xUnit and NUnit tests on the official Microsoft Learn page:
https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/vstest-v3?view=azure-pipelines#i-am-running-into-issues-when-running-data-driven-xunit-nunit-and-mstest-tests-with-some-of-the-task-options-are-there-known-limitations

We are working closely with the VSTest team to gain a deeper understanding of their design, which will help us explore viable options for supporting re-run functionality in data-driven tests. These insights will enable us to more accurately estimate the required effort and plan effectively.

In the short term, the recommended workaround is to convert data-driven tests to standard tests. This transition should be straightforward, and tools like GitHub Copilot can assist in making the process smoother.

We will proceed to close this issue for now, and future updates will be communicated through official blog posts.

**Comment by virekumar** [comment]
Hi Everyone,

Update here.
I did try a fix for "VsTest(Single Agent) Re-Run for data-driven in mstest, xunit and nunit with non displayName".

# **Nunit**

<img width="795" height="750" alt="Image" src="https://github.com/user-attachments/assets/2f30397f-7a20-46ff-96a5-503801a8a80f" />

# **Xunit** 

<img width="1598" height="750" alt="Image" src="https://github.com/user-attachments/assets/ca5ab2fa-ac6e-468a-8b2d-92a5d63be0c4" />

I will be merging and deploying these fixes incrementally and will keep on updating this thread.
As most of flows for data-driven are either not supportted or broken, so i will keep picking active issues for data-driven and will keep posting updates.

**Comment by virekumar** [comment]
Above fix is deployed. Please try it out for non-displayName and non-dynamic data-driven tests.

https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/vstest-v3?view=azure-pipelines#can-i-use-this-task-for-data-driven-parameterized-tests
